rfc: binary: Support broadcast source0 (#869)
oneapi-src/oneDNN: rfcs/20201103-binary-bcast-source0/README.md

# Proposal for enable broadcast source 0 in binary

## Problem and motivation
Currently in OneDNN library binary primitive supports implicit broadcast 

> semantics for source 1.It means that if some dimension has value of one, 

Reviewer: ```suggestion
semantics for source 1. It means that if some dimension has value of one, 
```

this value will be used to compute an operation with each point of 

> source 0 for this dimension. There is also a need for enable this functionality

Reviewer: ```suggestion
source 0 for this dimension. There is also a need for enabling this functionality
```

for source 1 to handle binary operation on sources like `Nx1` and `1xM`. This 

> will help better support OneDNN in TensorFlow library. At this point, was 

Reviewer: ```suggestion
will improve oneDNN integration in TensorFlow library. At this point, was 
```

> reported, that performance(jit kernel) is not the main concern.

Reviewer: ```suggestion
reported, that performance (jit kernel) is not the main concern.
```

![](bcast_example.png "Binary ADD operation 1x4:4x1")

## Overview

### Current implementations
*  Reference implementation allows broadcast with following syntax:
   `NxCxDxHxW:N/1xC/1xD/1xH/1xW/1 -> NxCxDxHxW`
*  JIT implementation allows broadcast with following syntax:
   `NxCxDxHxW:{NxCx1x1x1,1xCx1x1x1,1x1x1x1x1} -> NxCxDxHxW`
*  JIT i8 implementation allows broadcast with following syntax:
   `NxCxDxHxW:{N,1}xCx1x1x1 -> NxCxDxHxW`
### New implementations
*  Reference implementation allows broadcast with following syntax:
   `N/1xC/1xD/1xH/1xW/1:N/1xC/1xD/1xH/1xW/1 -> NxCxDxHxW`
### Possible future implementations (if the need is reported)
*  JIT implementation allows broadcast with following syntax:
   `{NxCxDxHxW,NxCx1x1x1,1xCx1x1x1,1x1x1x1x1}:
    {NxCxDxHxW,NxCx1x1x1,1xCx1x1x1,1x1x1x1x1} -> NxCxDxHxW`
*  JIT i8 implementation allows broadcast with following syntax:
   `{N,1}xCx{D,1}x{H,1}x{W,1}:{N,1}xCx{D,1}x{H,1}x{W,1} -> NxCxDxHxW`

Similarly for other formats.

## Proposal

> The proposal is to modify reference algorithm to enable any broadcast on 

Reviewer: Please rewrite into something like this:

> The proposal is to extend binary primitive by removing a restriction on src0 broadcasting. The reference implementation will be the only implementation supporting this broadcasting until we have an additional request.

Author: Done.

source 0 and source 1. The dimensions of the destination should be defined as: 
`dst_dims[d] = max(src0_dims[d], src1_dims[d])` and 
`src0_dims[d] == src1_dims[d] || one_of(1, src0_dims[d], src1_dims[d])`. Thanks 
to this, we will get backward compatibility. If user doesnt use broadcast on 
source0, source0 dimensions are the same as dst. The post-ops operations also 

> remains unchanged. If it will be reported, can also add support for broadcast 0

Reviewer: ```suggestion
remain unchanged. If it will be reported, can also add support for broadcast 0
```

> in jit kernels.

Reviewer: @bartekxk Thanks for the RFC!
In general proposal looks good, the only question is what to do with binary post operation? I think we should investigate this part deeper, because it might affect both developers and users in the future, so it should be at least documented. There are a few options:

1. Add a limitation to binary post operation so dst_dims should be equal to src0_dims.
    - Pros:
        - simple.
    - Cons:
         - complicates integration into an application. Before this change an application could match any combination of         some primitive and a binary primitive and fuse them together without looking into shapes. After this change an         application should track binary sizes and create a fusion only when src0_dims are equal to dst_dims.
2. Extend binary post operation to support src0 broadcast.
    - Pros:
        - ease of use (see cons of the first option).
    - Cons:
        - complicates post operations a lot:
            - post operation changes destination sizes of a primitive it is a part of.
            - I believe implementation will be very complicated due to broadcast of src0 and will require additional buffers which will limit performance benefit from binary post operation.

There are might be other options I am not aware of, so please don't take these options as **only** options.

Author: In general there are probably only two options - support or not. One more thing we can consider is for example jit or ref implementation broadcast source 0 in post-ops. I mentioned about it in second option and I think it can introduce a lot of complications in documentation. I added your options with some changes. Thanks.

# Proposal for enable broadcast source 0 in binary

## Problem and motivation
Currently in OneDNN library binary primitive supports implicit broadcast 
semantics for source 1. It means that if some dimension has value of one, 
this value will be used to compute an operation with each point of 
source 0 for this dimension. There is also a need for enabling this functionality
for source 1 to handle binary operation on sources like `Nx1` and `1xM`. This 
will improve oneDNN integration in TensorFlow library. At this point, was 
reported, that performance (jit kernel) is not the main concern.

![](bcast_example.png "Binary ADD operation 1x4:4x1")

## Overview

### Current implementations
*  Reference implementation allows broadcast with following syntax:
   `NxCxDxHxW:N/1xC/1xD/1xH/1xW/1 -> NxCxDxHxW`
*  JIT implementation allows broadcast with following syntax:
   `NxCxDxHxW:{NxCx1x1x1,1xCx1x1x1,1x1x1x1x1} -> NxCxDxHxW`
*  JIT i8 implementation allows broadcast with following syntax:

>    `NxCxDxHxW:{N,1}xCx1x1x1 -> NxCxDxHxW`

Reviewer 2: I believe int8 implementation is aligned in terms of broadcast support.

Author: You're right, thanks, fixed.

### New implementations
*  Reference implementation allows broadcast with following syntax:
   `N/1xC/1xD/1xH/1xW/1:N/1xC/1xD/1xH/1xW/1 -> NxCxDxHxW`
### Possible future implementations (if the need is reported)
*  JIT implementation allows broadcast with following syntax:
   `{NxCxDxHxW,NxCx1x1x1,1xCx1x1x1,1x1x1x1x1}:
    {NxCxDxHxW,NxCx1x1x1,1xCx1x1x1,1x1x1x1x1} -> NxCxDxHxW`
*  JIT i8 implementation allows broadcast with following syntax:
   `{N,1}xCx{D,1}x{H,1}x{W,1}:{N,1}xCx{D,1}x{H,1}x{W,1} -> NxCxDxHxW`

Similarly for other formats.

## Proposal
The proposal is to extend binary primitive by removing a restriction 
on src0 broadcasting. The reference implementation will be the only 
implementation supporting this broadcasting until we have an additional 
request. The dimensions of the destination should be defined as: 

> `dst_dims[d] = max(src0_dims[d], src1_dims[d])` and 

Reviewer 2: Right now dst_md is allowed to have format_kind::any which further is converted into src0 memory format. Since dimensions will not match, what's the plan to handle format_kind::any for dst in such case?

Author: I added section for this problem with two solutions.

`src0_dims[d] == src1_dims[d] || one_of(1, src0_dims[d], src1_dims[d])`. Thanks 
to this, we will get backward compatibility. If user doesnt use broadcast on 
source0, source0 dimensions are the same as dst. The post-ops operations also 

> remain unchanged. If it will be reported, can also add support for broadcast 0
> in jit kernels.

Reviewer 2: ```suggestion
remain unchanged. Optimized version of broadcast for src0 will be added upon request.
```

### Post-ops handle

We currently have two options on how to handle post-ops with this change:

> 1. Add a limitation to binary post operation so dst_dims should be equal 
>    to src0_dims. (probably should be chosen as the need is not requested 
>    by users and to avoid complications in post-ops)

Reviewer 2: ```suggestion
1. Restrict binary post-op src1 md (input) to have dimension values greater then dst_md of original operation. This means that broadcast operation for src0 in case of binary post-op is prohibited, but broadcast for src1 will still be supported as is.
```

    - Pros:
        - Very simple solution.
    - Cons:

>          - Complicates integration into an application. Before this change 

Reviewer 2: I don't understand this point. User was able to create a binary post-op md with any dimension values and at the stage of creating a primitive, the error would pop up since dimensions don't match. I don't see how the proposed behavior changes this scenario.

Reviewer: Let me answer to this question as i am the one who brought it to the discussion. The integration of DNNL into some frameworks (TF) is separated into 2 stages:
1. Graph-rewriter. During this stage graph nodes are merged according to the fusions supported by oneDNN. For example, Conv and Binary nodes could be merged into a single node "ConvBinary". If you look into the [integration](https://github.com/tensorflow/tensorflow/blob/8ba31084528f63e623fb5f58d9ee09b148088c6d/tensorflow/core/common_runtime/mkl_layout_pass.cc), it doesn't call oneDNN code and doesn't try to create primitives, it just merges nodes based on operation kinds.
2. Execution. During this stage node "ConvBinary" will create conv with binary fusion, execute it and destroy.

So my point is that adding restrictions on a fusion of binary based on sizes will bring complications on integration side, because integration should also manage these restrictions. Basically you can't assume anymore that if conv is supported, binary is supported, conv + binary fusion is supported from API perspective then conv + binary implementation will be supported.

         an application could match any combination of some primitive and a 
         binary primitive and fuse them together without looking into shapes. 
         After this change an application should track binary sizes and create 
         a fusion only when src0_dims are equal to dst_dims.

> 2. Extend binary post operation to support src0 broadcast.

Reviewer 2: Is there a real-world use case for such scenario at all?

Author: I don't know of any need for this use, and where it could be used.

    - Pros:
        - Ease of use (see cons of the first option).
    - Cons:

>         - It should be supported as jit version which is supported in most 

Reviewer 2: Just a side note: for current coverage of optimized binary post-op version, this is quite irrelevant. There are only two cases supported: a single point and a channel-wise. Single point will work regardless of dst_md dims values, but doing per channel broadcast in case dst_md has `C=1` doesn't make sense to me at all. So I won't add this one as a con.

Author: So why not add it as con?

    - Cons:
         - Complicates integration into an application. Before this change 
         an application could match any combination of some primitive and a 
         binary primitive and fuse them together without looking into shapes. 
         After this change an application should track binary sizes and create 
         a fusion only when src0_dims are equal to dst_dims.
2. Extend binary post operation to support src0 broadcast.
    - Pros:
        - Ease of use (see cons of the first option).
    - Cons:
        - It should be supported as jit version which is supported in most 
        primitives. The absence of this would introduce a lot of complications 
        in the documentation.
        - complicates post operations a lot:
            - post operation changes destination sizes of a primitive it is 
            a part of.
            - I believe implementation will be very complicated due to 
            broadcast of src0 and will require additional buffers which will 
            limit performance benefit from binary post operation.

### Any format handle

Right now dst_md is allowed to have format_kind::any which further is 
converted into src0 memory format.

> 1. Conver dst_md into max of dims from src0 and src1. Do it as it should be 
>    defined (preferred)

Reviewer 2: ```suggestion
1. (Recommended). Use the format of "bigger" tensor in a single broadcast case (currently, it's always src0, but it may be src1 since this feature lands). In case of double broadcasting, create a plain layout as of now, until requested something else.
```
